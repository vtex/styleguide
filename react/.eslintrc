{
  "extends": "vtex-react",
  "env": {
    "browser": true,
    "es6": true,
    "jest": true
  }
}
