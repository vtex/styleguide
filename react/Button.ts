import Button from './components/Button/index'

export default Button
