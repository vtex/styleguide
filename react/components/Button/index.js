import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

import Spinner from '../Spinner'
import { withForwardedRef, refShape } from '../../modules/withForwardedRef'

class Button extends Component {
  handleClick = event => {
    !this.props.disabled &&
      !this.props.isLoading &&
      this.props.onClick &&
      this.props.onClick(event)
  }

  componentDidMount() {
    if (this.props.icon) {
      console.warn(
        'Button: The prop "icon" of the "Button" component has been deprecated, and will be removed in a future version. Please use the component "ButtonWithIcon" instead'
      )
    }
    if (this.props.collapseLeft || this.props.collapseRight) {
      console.warn(
        'Button: The props "collapseLeft" and "collapseRight" of the "Button" component have been deprecated, and will be removed in a future version. Please use the component "ButtonPlain" instead'
      )
    }
  }

  render() {
    const {
      size,
      block,
      variation,
      icon,
      children,
      isLoading,
      collapseLeft,
      collapseRight,
      isGrouped,
      isActiveOfGroup,
      isFirstOfGroup,
      isLastOfGroup,
      href,
      target,
      rel,
      referrerPolicy,
      download,
      noUpperCase,
      noWrap,
      tabIndex,
    } = this.props

    const disabled = this.props.disabled || isLoading
    const iconOnly = icon || this.props.iconOnly

    const isTertiary =
      variation === 'tertiary' ||
      variation === 'inverted-tertiary' ||
      variation === 'danger-tertiary'

    let classes = 'vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid '
    let labelClasses =
      'vtex-button__label flex items-center justify-center h-100 '
    let loaderSize = 15
    let horizontalPadding = 0

    classes += !isGrouped ? 'br2 ' : ''
    classes += isFirstOfGroup ? 'br2 br--left ' : ''
    classes += isLastOfGroup ? 'br2 br--right ' : ''

    switch (size) {
      case 'small':
        classes += 'min-h-small t-action--small '
        horizontalPadding = iconOnly ? 3 : 5
        break
      case 'large':
        classes += 'min-h-large t-action--large '
        horizontalPadding = iconOnly ? 5 : 7
        loaderSize = 25
        break
      default:
        classes += 'min-h-regular t-action '
        horizontalPadding = iconOnly ? 4 : 6
        loaderSize = 20
        break
    }

    if (isTertiary && (collapseLeft || collapseRight)) {
      horizontalPadding = 2
    }

    labelClasses += `ph${horizontalPadding} `

    if (isTertiary && noUpperCase) {
      labelClasses += 'ttn '
    }

    if (collapseLeft && isTertiary) {
      labelClasses += `nl1 `
    }
    if (collapseRight && isTertiary) {
      labelClasses += `nr1 `
    }
    if (noWrap) {
      labelClasses += 'nowrap '
    }

    if (iconOnly) {
      classes += 'icon-button dib '
    }

    const primaryEnabledClasses =
      'bg-action-primary b--action-primary c-on-action-primary hover-bg-action-primary hover-b--action-primary hover-c-on-action-primary '
    const secondaryEnabledClasses =
      'bg-action-secondary b--action-secondary c-on-action-secondary hover-bg-action-secondary hover-b--action-secondary hover-c-on-action-secondary '

    if (isGrouped && !disabled) {
      classes += isActiveOfGroup
        ? primaryEnabledClasses
        : secondaryEnabledClasses
    } else {
      switch (variation) {
        default:
        case 'primary': {
          if (disabled) {
            classes += 'bg-disabled b--muted-5 c-on-disabled '
          } else {
            classes += primaryEnabledClasses
          }
          break
        }
        case 'secondary': {
          if (disabled) {
            classes += 'bg-disabled b--muted-5 c-on-disabled '
          } else {
            classes += secondaryEnabledClasses
          }
          break
        }
        case 'tertiary': {
          if (disabled) {
            classes += 'bg-transparent b--transparent c-disabled '
          } else {
            classes +=
              'bg-transparent b--transparent c-action-primary hover-b--transparent '
          }
          if (!disabled) {
            classes += 'hover-bg-action-secondary hover-b--action-secondary '
          }
          break
        }
        case 'inverted-tertiary': {
          if (disabled) {
            classes += 'bg-transparent b--transparent c-disabled '
          } else {
            classes += 'bg-transparent b--transparent c-on-base--inverted '
          }
          break
        }
        case 'danger': {
          if (disabled) {
            classes += 'bg-disabled b--muted-5 c-on-disabled '
          } else {
            classes +=
              'bg-danger b--danger c-on-danger hover-bg-danger hover-b--danger hover-c-on-danger '
          }
          break
        }
        case 'danger-tertiary': {
          if (disabled) {
            classes += 'bg-transparent b--transparent c-disabled '
          } else {
            classes +=
              'bg-transparent b--transparent c-danger hover-bg-danger--faded hover-b--danger--faded '
          }
          break
        }
      }
    }

    if (!disabled) {
      classes += 'pointer '
    }

    if (block) {
      classes += 'w-100 '
      labelClasses += 'w-100 border-box '
    }

    if (href) {
      classes += 'inline-flex items-center no-underline '
    }

    const style = {}

    if (iconOnly) {
      style.fontSize = 0
    }

    // Active state receives the hover color of the Button.
    // No token available for this.
    if (isActiveOfGroup && !disabled) {
      style.backgroundColor = '#0c389f'
      style.borderColor = '#0c389f'
      style.zIndex = 2
    }

    const linkModeProps = {
      target,
      rel,
      referrerPolicy,
      download,
    }

    const Element = href ? 'a' : 'button'

    return (
      <Element
        id={this.props.id}
        data-testid={this.props.testId}
        autoFocus={iconOnly ? undefined : this.props.autoFocus}
        disabled={iconOnly ? undefined : this.props.disabled}
        name={iconOnly ? undefined : this.props.name}
        value={iconOnly ? undefined : this.props.value}
        tabIndex={tabIndex}
        className={classes}
        href={href}
        onClick={this.handleClick}
        onMouseEnter={this.props.onMouseEnter}
        onMouseLeave={this.props.onMouseLeave}
        onMouseOver={this.props.onMouseOver}
        onMouseOut={this.props.onMouseOut}
        onMouseUp={this.props.onMouseUp}
        onMouseDown={this.props.onMouseDown}
        onFocus={this.props.onFocus}
        onBlur={this.props.onBlur}
        ref={this.props.forwardedRef}
        style={style}
        // Button-mode exclusive props
        type={href ? undefined : this.props.type}
        // Link-mode exclusive props
        {...(href && linkModeProps)}>
        {isLoading ? (
          <Fragment>
            <span className="vtex-button__spinner-container top-0 left-0 w-100 h-100 absolute flex justify-center items-center">
              <Spinner
                secondary={variation === 'primary' || variation === 'danger'}
                size={loaderSize}
              />
            </span>
            <span className={`${labelClasses} o-0`}>{children}</span>
          </Fragment>
        ) : (
          <div
            className={labelClasses}
            style={{
              paddingTop: '.25em',
              paddingBottom: '.32em',
            }}>
            {children}
          </div>
        )}
      </Element>
    )
  }
}

Button.defaultProps = {
  size: 'regular',
  block: false,
  variation: 'primary',
  disabled: false,
  autoFocus: false,
  icon: false,
  type: 'button',
  isLoading: false,
  isGrouped: false,
  isFirstOfGroup: false,
  isLastOfGroup: false,
  isActiveOfGroup: false,
  tabIndex: 0,
}

Button.propTypes = {
  /** Button size  */
  size: PropTypes.oneOf(['small', 'regular', 'large']),
  /** Button prominence variation */
  variation: PropTypes.oneOf([
    'primary',
    'secondary',
    'tertiary',
    'inverted-tertiary',
    'danger',
    'danger-tertiary',
  ]),
  /** Block style */
  block: PropTypes.bool,
  /** Loading state */
  isLoading: PropTypes.bool,
  /** [DEPRECATED] If you are using just an Icon component inside, use this as true */
  icon: PropTypes.bool,
  /** @ignore For internal use
   * Sets reduced paddings in order to keep the button squareish if it
   * only has an icon  */
  iconOnly: PropTypes.bool,
  /** (Button spec attribute) */
  id: PropTypes.string,
  /** Data attribute */
  testId: PropTypes.string,
  /** (Button spec attribute) */
  autoFocus: PropTypes.bool,
  /** (Button spec attribute) */
  autoComplete: PropTypes.string,
  /** (Button spec attribute) */
  disabled: PropTypes.bool,
  /** @ignore Forwarded Ref */
  forwardedRef: refShape,
  /** (Button spec attribute) */
  name: PropTypes.string,
  /** (Button spec attribute) */
  type: PropTypes.string,
  /** (Button spec attribute) */
  value: PropTypes.string,
  /** Label of the Button */
  children: PropTypes.node.isRequired,
  /** onClick event. */
  onClick: PropTypes.func,
  /** URL for link mode. Converts the button internally to a link. */
  href: PropTypes.string,
  /** onMouseEnter event */
  onMouseEnter: PropTypes.func,
  /** onMouseLeave event */
  onMouseLeave: PropTypes.func,
  /** onMouseOver event */
  onMouseOver: PropTypes.func,
  /** onMouseOut event */
  onMouseOut: PropTypes.func,
  /** onMouseUp event */
  onMouseUp: PropTypes.func,
  /** onMouseDown event */
  onMouseDown: PropTypes.func,
  /** onFocus event */
  onFocus: PropTypes.func,
  /** onBlur event */
  onBlur: PropTypes.func,
  /** tabIndex attribute of HTML */
  tabIndex: PropTypes.number,
  /** @ignore deprecated
   * Cancels out left padding */
  collapseLeft: PropTypes.bool,
  /** @ignore deprecated
   * Cancels out right padding */
  collapseRight: PropTypes.bool,
  /** */
  isGrouped: PropTypes.bool,
  /** */
  isFirstOfGroup: PropTypes.bool,
  /** */
  isLastOfGroup: PropTypes.bool,
  /** */
  isActiveOfGroup: PropTypes.bool,
  /** Link spec */
  target: PropTypes.string,
  /** Link spec */
  rel: PropTypes.string,
  /** Link spec */
  referrerPolicy: PropTypes.string,
  /** Link spec */
  download: PropTypes.string,
  /** When terciary, the upper case can be prevented */
  noUpperCase: PropTypes.bool,
  /** Disables label wrapping */
  noWrap: PropTypes.bool,
}

export default withForwardedRef(Button)
