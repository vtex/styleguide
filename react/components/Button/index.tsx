import React, {
  Fragment,
  SyntheticEvent,
  CSSProperties,
  useEffect,
  FC,
} from 'react'
import PropTypes, { InferProps } from 'prop-types'

import Spinner from '../Spinner'
import { withForwardedRef, refShape } from '../../modules/withForwardedRef'

const propTypes = {
  /** Button size  */
  size: PropTypes.oneOf(['small', 'regular', 'large']),
  /** Button prominence variation */
  variation: PropTypes.oneOf([
    'primary',
    'secondary',
    'tertiary',
    'inverted-tertiary',
    'danger',
    'danger-tertiary',
  ]),
  /** Block style */
  block: PropTypes.bool,
  /** Loading state */
  isLoading: PropTypes.bool,
  /** [DEPRECATED] If you are using just an Icon component inside, use this as true */
  icon: PropTypes.bool.isRequired,
  /** @ignore For internal use
   * Sets reduced paddings in order to keep the button squareish if it
   * only has an icon  */
  iconOnly: PropTypes.bool,
  /** (Button spec attribute) */
  id: PropTypes.string,
  /** Data attribute */
  testId: PropTypes.string,
  /** (Button spec attribute) */
  autoFocus: PropTypes.bool,
  /** (Button spec attribute) */
  autoComplete: PropTypes.string,
  /** (Button spec attribute) */
  disabled: PropTypes.bool.isRequired,
  /** @ignore Forwarded Ref */
  forwardedRef: refShape,
  /** (Button spec attribute) */
  name: PropTypes.string,
  /** (Button spec attribute) */
  type: PropTypes.string.isRequired,
  /** (Button spec attribute) */
  value: PropTypes.string,
  /** Label of the Button */
  children: PropTypes.node.isRequired,
  /** onClick event. */
  onClick: PropTypes.func,
  /** URL for link mode. Converts the button internally to a link. */
  href: PropTypes.string,
  /** onMouseEnter event */
  onMouseEnter: PropTypes.func,
  /** onMouseLeave event */
  onMouseLeave: PropTypes.func,
  /** onMouseOver event */
  onMouseOver: PropTypes.func,
  /** onMouseOut event */
  onMouseOut: PropTypes.func,
  /** onMouseUp event */
  onMouseUp: PropTypes.func,
  /** onMouseDown event */
  onMouseDown: PropTypes.func,
  /** onFocus event */
  onFocus: PropTypes.func,
  /** onBlur event */
  onBlur: PropTypes.func,
  /** @ignore deprecated
   * Cancels out left padding */
  collapseLeft: PropTypes.bool,
  /** @ignore deprecated
   * Cancels out right padding */
  collapseRight: PropTypes.bool,
  /** */
  isGrouped: PropTypes.bool,
  /** */
  isFirstOfGroup: PropTypes.bool,
  /** */
  isLastOfGroup: PropTypes.bool,
  /** */
  isActiveOfGroup: PropTypes.bool,
  /** Link spec */
  target: PropTypes.string,
  /** Link spec */
  rel: PropTypes.string,
  /** Link spec */
  referrerPolicy: PropTypes.string,
  /** Link spec */
  download: PropTypes.string,
  /** When terciary, the upper case can be prevented */
  noUpperCase: PropTypes.bool,
  /** Disables label wrapping */
  noWrap: PropTypes.bool,
}

type Props = InferProps<typeof propTypes>

const Button: FC<Props> = props => {
  const {
    id,
    size,
    block,
    variation,
    icon,
    children,
    isLoading,
    collapseLeft,
    collapseRight,
    isGrouped,
    isActiveOfGroup,
    isFirstOfGroup,
    isLastOfGroup,
    href,
    target,
    rel,
    referrerPolicy,
    download,
    noUpperCase,
    noWrap,
    testId,
    autoFocus,
    name,
    onMouseEnter,
    onMouseLeave,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onMouseDown,
    onFocus,
    onBlur,
    forwardedRef,
    type,
    value,
  } = props

  useEffect(() => {
    if (icon) {
      console.warn(
        'Button: The prop "icon" of the "Button" component has been deprecated, and will be removed in a future version. Please use the component "ButtonWithIcon" instead'
      )
    }
    if (collapseLeft || collapseRight) {
      console.warn(
        'Button: The props "collapseLeft" and "collapseRight" of the "Button" component have been deprecated, and will be removed in a future version. Please use the component "ButtonPlain" instead'
      )
    }
  }, [icon, collapseLeft, collapseRight])

  const handleClick = (event: SyntheticEvent) => {
    if (!props.disabled && !props.isLoading && props.onClick) {
      props.onClick(event)
    }
  }

  const disabled = props.disabled || isLoading
  const iconOnly = icon || props.iconOnly

  const isTertiary =
    variation === 'tertiary' ||
    variation === 'inverted-tertiary' ||
    variation === 'danger-tertiary'

  let classes = 'vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid '
  let labelClasses =
    'vtex-button__label flex items-center justify-center h-100 '
  let loaderSize = 15
  let horizontalPadding = 0

  classes += !isGrouped ? 'br2 ' : ''
  classes += isFirstOfGroup ? 'br2 br--left ' : ''
  classes += isLastOfGroup ? 'br2 br--right ' : ''

  switch (size) {
    case 'small':
      classes += 'min-h-small t-action--small '
      horizontalPadding = iconOnly ? 3 : 5
      break
    case 'large':
      classes += 'min-h-large t-action--large '
      horizontalPadding = iconOnly ? 5 : 7
      loaderSize = 25
      break
    default:
      classes += 'min-h-regular t-action '
      horizontalPadding = iconOnly ? 4 : 6
      loaderSize = 20
      break
  }

  if (isTertiary && (collapseLeft || collapseRight)) {
    horizontalPadding = 2
  }

  labelClasses += `ph${horizontalPadding} `

  if (isTertiary && noUpperCase) {
    labelClasses += 'ttn '
  }

  if (collapseLeft && isTertiary) {
    labelClasses += `nl1 `
  }
  if (collapseRight && isTertiary) {
    labelClasses += `nr1 `
  }
  if (noWrap) {
    labelClasses += 'nowrap '
  }

  if (iconOnly) {
    classes += 'icon-button dib '
  }

  const primaryEnabledClasses =
    'bg-action-primary b--action-primary c-on-action-primary hover-bg-action-primary hover-b--action-primary hover-c-on-action-primary '
  const secondaryEnabledClasses =
    'bg-action-secondary b--action-secondary c-on-action-secondary hover-bg-action-secondary hover-b--action-secondary hover-c-on-action-secondary '

  if (isGrouped && !disabled) {
    classes += isActiveOfGroup ? primaryEnabledClasses : secondaryEnabledClasses
  } else {
    switch (variation) {
      default:
      case 'primary': {
        if (disabled) {
          classes += 'bg-disabled b--muted-5 c-on-disabled '
        } else {
          classes += primaryEnabledClasses
        }
        break
      }
      case 'secondary': {
        if (disabled) {
          classes += 'bg-disabled b--muted-5 c-on-disabled '
        } else {
          classes += secondaryEnabledClasses
        }
        break
      }
      case 'tertiary': {
        if (disabled) {
          classes += 'bg-transparent b--transparent c-disabled '
        } else {
          classes +=
            'bg-transparent b--transparent c-action-primary hover-b--transparent '
        }
        if (!disabled) {
          classes += 'hover-bg-action-secondary hover-b--action-secondary '
        }
        break
      }
      case 'inverted-tertiary': {
        if (disabled) {
          classes += 'bg-transparent b--transparent c-disabled '
        } else {
          classes += 'bg-transparent b--transparent c-on-base--inverted '
        }
        break
      }
      case 'danger': {
        if (disabled) {
          classes += 'bg-disabled b--muted-5 c-on-disabled '
        } else {
          classes +=
            'bg-danger b--danger c-on-danger hover-bg-danger hover-b--danger hover-c-on-danger '
        }
        break
      }
      case 'danger-tertiary': {
        if (disabled) {
          classes += 'bg-transparent b--transparent c-disabled '
        } else {
          classes +=
            'bg-transparent b--transparent c-danger hover-bg-danger--faded hover-b--danger--faded '
        }
        break
      }
    }
  }

  if (!disabled) {
    classes += 'pointer '
  }

  if (block) {
    classes += 'w-100 '
    labelClasses += 'w-100 border-box '
  }

  if (href) {
    classes += 'inline-flex items-center no-underline '
  }

  const style: CSSProperties = {}

  if (iconOnly) {
    style.fontSize = 0
  }

  // Active state receives the hover color of the Button.
  // No token available for this.
  if (isActiveOfGroup && !disabled) {
    style.backgroundColor = '#0c389f'
    style.borderColor = '#0c389f'
  }

  const linkModeProps = {
    target,
    rel,
    referrerPolicy,
    download,
  }

  const Element: React.ElementType = href ? 'a' : 'button'

  const elementDefaultProps = {
    id,
    'data-testid': testId,
    tabIndex: 0,
    className: classes,
    onClick: handleClick,
    onMouseEnter,
    onMouseLeave,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onMouseDown,
    onFocus,
    onBlur,
    ref: forwardedRef,
    style,
  }

  let elementEspecificProps
  if (href) {
    // Link-mode exclusive props
    elementEspecificProps = {
      ...linkModeProps,
      href,
    }
  } else {
    // Button-mode exclusive props
    elementEspecificProps = {
      autoFocus: iconOnly ? undefined : autoFocus,
      disabled: iconOnly ? undefined : props.disabled,
      name: iconOnly ? undefined : name,
      value: iconOnly ? undefined : value,
      type,
    }
  }

  return (
    <Element {...elementDefaultProps} {...elementEspecificProps}>
      {isLoading ? (
        <Fragment>
          <span className="vtex-button__spinner-container top-0 left-0 w-100 h-100 absolute flex justify-center items-center">
            <Spinner
              secondary={variation === 'primary' || variation === 'danger'}
              size={loaderSize}
            />
          </span>
          <span className={`${labelClasses} o-0`}>{children}</span>
        </Fragment>
      ) : (
        <div
          className={labelClasses}
          style={{
            paddingTop: '.25em',
            paddingBottom: '.32em',
          }}
        >
          {children}
        </div>
      )}
    </Element>
  )
}

Button.propTypes = propTypes
Button.defaultProps = {
  size: 'regular',
  block: false,
  variation: 'primary',
  disabled: false,
  autoFocus: false,
  icon: false,
  type: 'button',
  isLoading: false,
  isGrouped: false,
  isFirstOfGroup: false,
  isLastOfGroup: false,
  isActiveOfGroup: false,
}

export default withForwardedRef<HTMLButtonElement & HTMLAnchorElement, Props>(
  Button
)
