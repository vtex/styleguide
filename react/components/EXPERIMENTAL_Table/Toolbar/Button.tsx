import React, { forwardRef, ReactNode } from 'react'
import classNames from 'classnames'

import ButtonWithIcon from '../../ButtonWithIcon/index.js'
import { E2ETestable } from '../types'

const ICON_OPTICAL_COMPENSATION = { marginTop: 1.5 }

type Ref = HTMLDivElement

const Button = forwardRef<Ref, ButtonProps>(
  (
    {
      id,
      title,
      disabled,
      onClick,
      icon,
      isLoading,
      children,
      label,
      variation,
      isActiveOfGroup,
      isGrouped,
      isFirstOfGroup,
      size,
      testId,
    },
    ref
  ) => {
    const isTertiary = variation === ButtonVariation.Tertiary

    const iconClass = classNames('mh2', {
      'c-on-base': isTertiary && !disabled,
      'c-muted-2': disabled,
    })

    const labelClass = classNames({
      'c-on-base': isTertiary && !disabled,
      'c-muted-2': disabled,
    })

    return (
      <div
        id={`${id}__container`}
        data-testid={`${testId}__container`}
        title={title}
        ref={ref}
        className={classNames('relative', { mh2: isTertiary })}
      >
        <ButtonWithIcon
          icon={
            <span className={iconClass} style={ICON_OPTICAL_COMPENSATION}>
              {icon}
            </span>
          }
          testId={testId}
          isGrouped={isGrouped}
          isFirstOfGroup={isFirstOfGroup}
          isActiveOfGroup={isActiveOfGroup}
          disabled={disabled}
          isLoading={isLoading}
          variation={variation}
          size={size}
          onClick={onClick}
        >
          {label && <span className={labelClass}>{label}</span>}
        </ButtonWithIcon>
        {children}
      </div>
    )
  }
)

export enum ButtonVariation {
  Primary = 'primary',
  Secondary = 'secondary',
  Tertiary = 'tertiary',
}

export enum ButtonSize {
  Small = 'small',
  Regular = 'regular',
  Large = 'large',
}

export enum IconSize {
  Heavy = 13,
  Medium = 14,
  Light = 16,
}

export type ButtonProps = E2ETestable & {
  id?: string
  label?: string
  onClick?: Function
  isLoading?: boolean
  disabled?: boolean
  size?: ButtonSize
  icon?: unknown
  title?: string
  variation?: ButtonVariation
  isActiveOfGroup?: boolean
  isGrouped?: boolean
  isFirstOfGroup?: boolean
  children?: ReactNode
}

Button.defaultProps = {
  variation: ButtonVariation.Tertiary,
  isActiveOfGroup: false,
  isGrouped: false,
  isFirstOfGroup: false,
  size: ButtonSize.Small,
}

export default Button
