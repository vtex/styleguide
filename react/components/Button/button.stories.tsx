import React from 'react'
import { withA11y } from '@storybook/addon-a11y'
import { withKnobs, text, boolean, select } from "@storybook/addon-knobs"
import { action } from '@storybook/addon-actions'

import Button from '.'

export default {
  title: 'Components|Button',
  component: Button,
  decorators: [withA11y, withKnobs]
}

const variations = ['primary', 'secondary', 'tertiary', 'danger', 'danger-tertiary', 'inverted-tertiary']

export const Default = () => (
  <Button
    variation={select('Variation', variations, 'primary')}
    isLoading={boolean('Is loading', false)}
    disabled={boolean('Disabled', false)}
    block={boolean('Block', false)}
    size={select('Size', ['small', 'regular', 'large'], 'regular')}
    onClick={action('button-click')}
    onFocus={action('button-focus')}
  >
    {text("Label", "With a text")}
  </Button>
)

export const withLoadingState = () => (
  <Button
    isLoading={true}
  >
    Loading
  </Button>
)

export const withDiffentVariations = () => (
  <>
    {variations.map(variation => (
      <>
        <Button variation={variation}>{variation}</Button>
        <br/>
        <br/>
      </>
    ))}
  </>
)

export const withOnMouseEvents = () => (
  <Button
    onMouseDown={action('button-on-mouse-down')}
    onMouseEnter={action('button-on-mouse-enter')}
    onMouseOver={action('button-on-mouse-over')}
    onMouseLeave={action('button-on-mouse-leave')}
    onMouseOut={action('button-on-mouse-out')}
    onMouseUp={action('button-on-mouse-up')}
  >
    Button
  </Button>
)
