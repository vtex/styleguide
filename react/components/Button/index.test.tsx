import React from 'react'
import { render } from '@testing-library/react'

import Button from '.'

describe('Button', () => {
  describe('CSS API', () => {
    it('primary', () => {
      const { asFragment } = render(<Button variation="primary">Hello</Button>)

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('secondary', () => {
      const { asFragment } = render(
        <Button variation="secondary">Hello</Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('tertiary', () => {
      const { asFragment } = render(<Button variation="tertiary">Hello</Button>)

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('danger', () => {
      const { asFragment } = render(<Button variation="danger">Hello</Button>)

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('danger-tertiary', () => {
      const { asFragment } = render(
        <Button variation="danger-tertiary">Hello</Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('inverted-tertiary', () => {
      const { asFragment } = render(
        <Button variation="inverted-tertiary">Hello</Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('primary disabled', () => {
      const { asFragment } = render(
        <Button variation="primary" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('secondary disabled', () => {
      const { asFragment } = render(
        <Button variation="secondary" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('tertiary disabled', () => {
      const { asFragment } = render(
        <Button variation="tertiary" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('danger disabled', () => {
      const { asFragment } = render(
        <Button variation="danger" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('danger-tertiary disabled', () => {
      const { asFragment } = render(
        <Button variation="danger-tertiary" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })

    it('inverted-tertiary disabled', () => {
      const { asFragment } = render(
        <Button variation="inverted-tertiary" disabled>
          Hello
        </Button>
      )

      const result = asFragment()

      expect(result).toMatchSnapshot()
    })
  })
})
