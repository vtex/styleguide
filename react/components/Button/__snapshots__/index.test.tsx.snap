// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button CSS API danger 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-danger b--danger c-on-danger hover-bg-danger hover-b--danger hover-c-on-danger pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API danger disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-disabled b--muted-5 c-on-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API danger-tertiary 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-danger hover-bg-danger--faded hover-b--danger--faded pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API danger-tertiary disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API inverted-tertiary 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-on-base--inverted pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API inverted-tertiary disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API primary 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-action-primary b--action-primary c-on-action-primary hover-bg-action-primary hover-b--action-primary hover-c-on-action-primary pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API primary disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-disabled b--muted-5 c-on-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API secondary 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-action-secondary b--action-secondary c-on-action-secondary hover-bg-action-secondary hover-b--action-secondary hover-c-on-action-secondary pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API secondary disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-disabled b--muted-5 c-on-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API tertiary 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-action-primary hover-b--transparent hover-bg-action-secondary hover-b--action-secondary pointer "
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button CSS API tertiary disabled 1`] = `
<DocumentFragment>
  <button
    class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-regular t-action bg-transparent b--transparent c-disabled "
    disabled=""
    tabindex="0"
    type="button"
  >
    <div
      class="vtex-button__label flex items-center justify-center h-100 ph6 "
      style="padding-top: .25em; padding-bottom: .32em;"
    >
      Hello
    </div>
  </button>
</DocumentFragment>
`;
